### PR TITLE
ddtrace/tracer: sample child spans with no priority

### PR DESCRIPTION
**Pull request by Jesse-Cameron**
fixes: #793

Sample the span via config if the parent span ctx does not contain the necessary sampling priority details.

I had a little through getting the `t.sample` to work for the test cases. So I moved the check down until after span has a default context initialized. 

BTW: I had some trouble getting the tests to work in my instance of circleci due to plan limitations. Would appreciate some help running in the datadog instance if possible :)

### CODE REVIEW

**Review by knusbaum — CHANGES_REQUESTED** [review]
Thanks for this! 

I think this is close. 

The `StartSpan` logic is a little convoluted. For instance, we don't set the span's context until after [this block](https://github.com/DataDog/dd-trace-go/blob/adb865a21a52f60f10d77ca3e6cf4e2f5b8db4e2/ddtrace/tracer/tracer.go#L269-L288). I think that is because [`newSpanContext` uses `span.SpanID` and `span.TraceID` to populate the context's values.](https://github.com/DataDog/dd-trace-go/blob/adb865a21a52f60f10d77ca3e6cf4e2f5b8db4e2/ddtrace/tracer/spancontext.go#L47-L51) It's a little confusing because in the code block right before we call `newSpanContext`, we set the [span id and trace id from the parent context](https://github.com/DataDog/dd-trace-go/blob/adb865a21a52f60f10d77ca3e6cf4e2f5b8db4e2/ddtrace/tracer/tracer.go#L271-L272) anyway.

> BTW: I had some trouble getting the tests to work in my instance of circleci due to plan limitations. Would appreciate some help running in the datadog instance if possible :)

You are free to push commits to this PR, and CircleCI will run. You are also free to put up a draft PR at any time where tests will run. 

Another option (and the easiest, in my opinion) is to run the tracer tests locally. This is more difficult for `contrib/*`, but since you're working on `ddtrace/tracer`, you should be able to `cd` into the `ddtrace/tracer` directory and run `go test` to run all the core tests.

**Comment by Jesse-Cameron** [comment]
Hey @knusbaum ,

I will admit, I found the `StartSpan` logic a little confusing myself.

> The StartSpan logic is a little convoluted. For instance, we don't set the span's context until after this block. I think that is because newSpanContext uses span.SpanID and span.TraceID to populate the context's values. It's a little confusing because in the code block right before we call newSpanContext, we set the span id and trace id from the parent context anyway.

I'm not sure that we can set the `samplingPriority` inside of the first `if context != nil` block. I did originally try this. But the `context.samplingPriority()` call requires the context to have a `trace`. And that `trace` to have a `samplingPriority`. Which keep causing the tests to fail.

When why I elected to try and do the sampling after the `newSpanContext`.

> You are free to push commits to this PR, and CircleCI will run. You are also free to put up a draft PR at any time where tests will run.

Weird, when I push to this branch it seems to be running under my plan, not the datadog one. Am I doing something wrong?

File: ddtrace/tracer/tracer.go [hunk 1]
```diff
@@ -318,6 +318,10 @@ func (t *tracer) StartSpan(operationName string, options ...ddtrace.StartSpanOpt
 	if t.config.env != "" {
 		span.SetTag(ext.Environment, t.env)
 	}
+	if _, ok := span.context.samplingPriority(); !ok {
+		// if not already sampled
+		t.sample(span)
+	}
 	if context == nil {
 		// this is a brand new trace, sample it
 		t.sample(span)
```
[the remark this answers is not in the document]
Comment (Jesse-Cameron) [reply ddtrace/tracer/tracer.go]: Yep, I think I know what you are saying. And I've made that change in 4377471.

File: ddtrace/tracer/tracer.go [hunk 2]
```diff
@@ -351,9 +355,11 @@ const sampleRateMetricKey = "_sample_rate"
 
 // Sample samples a span with the internal sampler.
 func (t *tracer) sample(span *span) {
-	if _, ok := span.context.samplingPriority(); ok {
-		// sampling decision was already made
-		return
```
[the remark this answers is not in the document]
Comment (Jesse-Cameron) [reply ddtrace/tracer/tracer.go:356]: Yep, this was left in by mistake. Have removed

**Comment by Jesse-Cameron** [comment]
Hey @knusbaum , sorry to prod, but have you had a moment to look over this one? 😅

**Review by knusbaum — APPROVED** [review]
@Jesse-Cameron Thanks for your patience. I was out on leave.

I think this is a very clean solution to the issue. Thanks for working on this with us.

**Comment by knusbaum** [comment]
Figuring out the CircleCI thing. Very strange it's running under your account.

**Comment by Jesse-Cameron** [comment]
Np @knusbaum , everyone deserves a break!

> Figuring out the CircleCI thing. Very strange it's running under your account.

yeah I found this article: https://support.circleci.com/hc/en-us/articles/360008097173-Why-aren-t-pull-requests-triggering-jobs-on-my-organization-

I have unfollowed my fork. Perhaps, we can close this PR and open it again?

**Comment by Jesse-Cameron** [comment]
Ah okay. That seems to have fixed it. The failing test is due to the tagging:
image

**Review by knusbaum — APPROVED** [review]
LGTM

**Pull request merged**